### PR TITLE
refactor snapshot to use shared infromer and versioned client

### LINKED ISSUE #644: refactor snapshot to use shared infromer and versioned client
Snapshot uses CRD to define resources such as `VolumeSnapshot` and  `VolumeSnapshotData` .  We should follow the pattern https://github.com/kubernetes/sample-controller to implement our code and make it more readable and clear.

### PR DESCRIPTION
**Pull request by wackxu**
fix https://github.com/kubernetes-incubator/external-storage/issues/644

/assign @rootfs @tsmetana @jsafrane

### CODE REVIEW

**Comment by wackxu** [comment]
@rootfs @tsmetana @jsafrane PTAL

File: hack/update-generated-code.sh [hunk 1]
```diff
@@ -14,18 +14,21 @@
...
+#                  instead of the $GOPATH directly. For normal projects this can be dropped.
+${CODEGEN_PKG}/generate-groups.sh "deepcopy,client,informer,lister" \
+  github.com/kubernetes-incubator/external-storage/snapshot/pkg/client github.com/kubernetes-incubator/external-storage/snapshot/pkg/apis \
+  volumesnapshot:v1 \
+  --go-header-file ${SCRIPT_ROOT}/hack/boilerplate.go.txt
 
-echo "Generating deepcopy funcs"
-${OUTPUT_HOSTBIN}/deepcopy-gen --input-dirs ${ALL_FQ_APIS} -O zz_generated.deepcopy --bounding-dirs $ALL_FQ_APIS --logtostderr
+# To use your own boilerplate text use:
+#   --go-header-file ${SCRIPT_ROOT}/hack/custom-boilerplate.go.txt
```
Comment (tsmetana) [inline hack/update-generated-code.sh]: I failed to run hack/update-generated-code.sh successfully. It spits this and ends with error:
```
# github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/informer-gen/args
/home/tsmetana/go/src/github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/informer-gen/args/args.go:38:31: args.Default().WithoutDefaultFlagParsing undefined (type *args.GeneratorArgs has no field or method WithoutDefaultFlagParsing)
# github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/defaulter-gen/args
/home/tsmetana/go/src/github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/defaulter-gen/args/args.go:32:31: args.Default().WithoutDefaultFlagParsing undefined (type *args.GeneratorArgs has no field or method WithoutDefaultFlagParsing)
# github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/client-gen/args
/home/tsmetana/go/src/github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/client-gen/args/args.go:55:31: args.Default().WithoutDefaultFlagParsing undefined (type *args.GeneratorArgs has no field or method WithoutDefaultFlagParsing)
# github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/lister-gen/args
/home/tsmetana/go/src/github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/lister-gen/args/args.go:33:31: args.Default().WithoutDefaultFlagParsing undefined (type *args.GeneratorArgs has no field or method WithoutDefaultFlagParsing)
# github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/deepcopy-gen/args
/home/tsmetana/go/src/github.com/kubernetes-incubator/external-storage/vendor/k8s.io/code-generator/cmd/deepcopy-gen/args/args.go:32:31: args.Default().WithoutDefaultFlagParsing undefined (type *args.GeneratorArgs has no field or method WithoutDefaultFlagParsing)
```
Not sure what exactly is wrong but the code generation definitely works in the master branch... Obviously the resulting code does not work...

File: hack/verify-codegen.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ret=0
+diff -Naupr "${DIFFROOT}" "${TMP_DIFFROOT}" || ret=$?
+cp -a "${TMP_DIFFROOT}"/* "${DIFFROOT}"
+if [[ $ret -eq 0 ]]
+then
+  echo "${DIFFROOT} up to date."
+else
+  echo "${DIFFROOT} is out of date. Please run hack/update-codegen.sh"
+  exit 1
+fi
```
Comment (tsmetana) [inline hack/verify-codegen.sh]: Again: there are already various verify-* scripts in the tree (see the boilerplate.go.txt comment).

**Review by tsmetana — CHANGES_REQUESTED** [review]
The new code generation and verification issues should be fixed first.

**Pull request closed without merging**